Source/Core/Core/HLE/HLE_OS.cpp

```diff
@@ -75,14 +88,79 @@ void HLE_write_console()
   NOTICE_LOG(OSREPORT, "%08x->%08x| %s", LR, PC, SHIFTJISToUTF8(report_message).c_str());
 }
 
-std::string GetStringVA(u32 strReg)
+// Log (v)dprintf message if fd is 1 (stdout) or 2 (stderr)
+void HLE_LogDPrint(bool has_va_list)
 {
-  std::string ArgumentBuffer;
-  u32 ParameterCounter = strReg + 1;
-  u32 FloatingParameterCounter = 1;
+  NPC = LR;
+
+  if (GPR(3) != 1 && GPR(3) != 2)
+    return;
+
+  std::string report_message = GetStringVA(4, has_va_list);
+  NOTICE_LOG(OSREPORT, "%08x->%08x| %s", LR, PC, SHIFTJISToUTF8(report_message).c_str());
+}
+
+// Log dprintf message
+//  -> int dprintf(int fd, const char* format, ...);
+void HLE_LogDPrint()
+{
+  HLE_LogDPrint(false);
+}
+
+// Log vdprintf message
+//  -> int vdprintf(int fd, const char* format, va_list ap);
+void HLE_LogVDPrint()
+{
+  HLE_LogDPrint(true);
+}
+
+// Log (v)fprintf message if FILE is stdout or stderr
+void HLE_LogFPrint(bool has_va_list)
+{
+  NPC = LR;
+
+  // The structure FILE is implementation defined.
+  // Both libogc and Dolphin SDK seem to store the fd at the same address.
+  int fd = -1;
+  if (PowerPC::HostIsRAMAddress(GPR(3)) && PowerPC::HostIsRAMAddress(GPR(3) + 0xF))
+  {
+    // The fd is stored as a short at FILE+0xE.
+    fd = static_cast<short>(PowerPC::HostRead_U16(GPR(3) + 0xE));
+  }
+  if (fd != 1 && fd != 2)
+  {
+    // On RVL SDK it seems stored at FILE+0x2.
+    fd = static_cast<short>(PowerPC::HostRead_U16(GPR(3) + 0x2));
+  }
+  if (fd != 1 && fd != 2)
+    return;
 
+  std::string report_message = GetStringVA(4, has_va_list);
+  NOTICE_LOG(OSREPORT, "%08x->%08x| %s", LR, PC, SHIFTJISToUTF8(report_message).c_str());
+}
+
+// Log fprintf message
+//  -> int fprintf(FILE* stream, const char* format, ...);
+void HLE_LogFPrint()
+{
+  HLE_LogFPrint(false);
+}
+
+// Log vfprintf message
+//  -> int vfprintf(FILE* stream, const char* format, va_list ap);
+void HLE_LogVFPrint()
+{
+  HLE_LogFPrint(true);
+}
+
+std::string GetStringVA(u32 str_reg, bool has_va_list)
+{
+  std::string ArgumentBuffer;
   std::string result;
-  std::string string = PowerPC::HostGetString(GPR(strReg));
+  std::string string = PowerPC::HostGetString(GPR(str_reg));
+  std::unique_ptr<HLE::SystemVABI::VAList> ap =
```

Comment:
`auto ap`. I don't think there is a need to write HLE::SystemVABI::VAList again, considering it's already in the make_unique.